x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useMemo } from 'react';
+import { EuiButtonEmpty, EuiCallOut, EuiFormRow, EuiSpacer, EuiText } from '@elastic/eui';
+import type { DataViewFieldBase } from '@kbn/es-query';
+import type { RequiredFieldInput } from '../../../../../common/api/detection_engine';
+import { UseArray, useFormData } from '../../../../shared_imports';
+import type { FormHook, ArrayItem } from '../../../../shared_imports';
+import { RequiredFieldsHelpInfo } from './required_fields_help_info';
+import { RequiredFieldRow } from './required_fields_row';
+import * as i18n from './translations';
+
+interface RequiredFieldsComponentProps {
+  path: string;
+  indexPatternFields?: DataViewFieldBase[];
+  isIndexPatternLoading?: boolean;
+}
+
+const RequiredFieldsComponent = ({
+  path,
+  indexPatternFields = [],
+  isIndexPatternLoading = false,
+}: RequiredFieldsComponentProps) => {
+  return (
+    <UseArray path={path} initialNumberOfItems={0}>
+      {({ items, addItem, removeItem, form }) => (
+        <RequiredFieldsList
+          items={items}
+          addItem={addItem}
+          removeItem={removeItem}
+          indexPatternFields={indexPatternFields}
+          isIndexPatternLoading={isIndexPatternLoading}
+          path={path}
+          form={form}
+        />
+      )}
+    </UseArray>
+  );
+};
+
+interface RequiredFieldsListProps {
+  items: ArrayItem[];
+  addItem: () => void;
+  removeItem: (id: number) => void;
+  indexPatternFields: DataViewFieldBase[];
+  isIndexPatternLoading: boolean;
+  path: string;
+  form: FormHook;
+}
+
+const RequiredFieldsList = ({
+  items,
+  addItem,
+  removeItem,
+  indexPatternFields,
+  isIndexPatternLoading,
+  path,
+  form,
+}: RequiredFieldsListProps) => {
+  /*
+    This component should only re-render when either the "index" form field (index patterns) or the required fields change. 
+
+    By default, the `useFormData` hook triggers a re-render whenever any form field changes.
+    It also allows optimization by passing a "watch" array of field names. The component then only re-renders when these specified fields change.
+
+    Hovewer, it doesn't work with fields created using the `UseArray` component.
+    In `useFormData`, these array fields are stored as "flattened" objects with numbered keys, like { "requiredFields[0]": { ... }, "requiredFields[1]": { ... } }.
+    The "watch" feature of `useFormData` only works if you pass these "flattened" field names, such as ["requiredFields[0]", "requiredFields[1]", ...], not just "requiredFields".
```

Comment:
I wonder, what if you instead of getting all fields, converting it to array with keys and the parsing each key, just reach out directly to `requiredFields__array__` field?

<img width="565" alt="Screenshot 2024-05-15 at 11 29 49" src="https://github.com/elastic/kibana/assets/92328789/166c7788-6b57-4ccb-b19f-ffdd13d012f0">

Looks like it has in value property all the keys you need to watch